Scaffolder: `concurrentTasksLimit` for TaskWorker

howlowck opened the pull request:
## Enable a single TaskWorker to run multiple concurrent tasks

This is the implementation of #13799 RFC, where the conclusion is to allow a single TaskWorker to run multiple concurrent tasks up to a limit.

Changes:
* Added `concurrentTasksLimit` option to TaskWorker `create` static method.

A screenshot of two long-running tasks running in parallel with a single TaskWorker:
<img width="902" alt="image" src="https://user-images.githubusercontent.com/338265/193891847-6fb60bd2-b8ad-4c45-a1af-07a0f75ec0de.png">

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

freben commented:
as for the vale error, just put the offending "techy" words in back ticks and they'll be ignored by the spell checker

Rugvip commented on plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:
```diff
@@ -89,6 +94,15 @@ export class TaskWorker {
   }
 
+  listen() {
+    this.options.eventEmitter.on('storageTaskBroker:newTask', () => {
+      this.options.taskBroker.claim().then(task => {
+        this.runOneTask(task);
+      });
+    });
```
Tbh, I don't think we actually need the event emitter to achieve the non-blocking nature that you're looking for?

If you look at the implementation of the `start` method just above, the `claim()` method will return as soon as a new task is available. It's basically an event listener but one that's driven from the consumer end.

What makes the current implementation blocking is that we wait for `runOneTask` to complete. We do this so that we don't get a runaway effect where the backend gets overloaded due to too many tasks running at the same time. What we could do is to let multiple tasks run concurrently, but add a limit instead. For example, it might pick up and execute tasks up until there are 5 tasks in flight, at which point the implementation waits for one of them to complete before it calls `claim()` again.

howlowck replied:
Hi @Rugvip, thanks for your feedback! @benjdlambert also commented about the eventEmitter in the RFC. I responded to the point in [the comment of #13799](https://github.com/backstage/backstage/issues/13799#issuecomment-1256588769).

Can you talk a little more about the "runaway effect"? My interpretation of "overloading" the system is that the tasks will take up the memory. To me, it's the JS engine's job of allocating memory and managing the timing of the execution with the event loop. One badly placed infinite loop in one task will kill the runtime over 1000 concurrent fetch calls. So, restricting the number of concurrent tasks doesn't prevent overloading the server.

Rugvip replied:
It's based on the assumption that there is some amount of CPU or local I/O bound actions that need to run, and that it's better to queue them up instead of running them all at once. To be clear, I'm thinking about an order of 10-100 concurrent task executions.

Either way it doesn't matter much tbh, I think some form of maximum concurrent tasks configuration would make sense. But we can allow that to be set to a very high value, or infinity

howlowck replied:
Hi @Rugvip, as the result of the RFC. I've significantly simplified this PR. I'm going to resolve this comment thread. Thanks!

Rugvip reviewed (changes requested):
Thank you! 👍 

Some thoughts on how we achieve the non-blockiness

jhaals commented on plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:
```diff
@@ -83,14 +118,17 @@ export class TaskWorker {
   start() {
     (async () => {
       for (;;) {
+        await this.waitWorkerToBeAvailable();
```
Could we use something pre-existing like [p-queue](https://github.com/sindresorhus/p-queue), plimit or alike so that we don't have to maintain our own locking.

howlowck replied:
Awesome. Thanks for the suggestions. Totally agree about using a community-backed solution. I briefly looked into p-limit, I don't think we can used that, because it needs the array of promises before the creation of the limiting mechanism. `p-queue` looks to be more inline with the existing software architecture. I can look more into implementing it this week.

howlowck replied:
I implemented a taskQueue using the p-queue library instead of creating own locking mechanism

jhaals commented on plugins/scaffolder-backend/src/service/router.ts:
```diff
@@ -202,7 +202,7 @@ export async function createRouter(
 
   const workers = [];
-  for (let i = 0; i < (taskWorkers || 3); i++) {
+  for (let i = 0; i < (taskWorkers || 1); i++) {
```
We probably want to deprecate this option in favour of `concurrentTasksLimit`

howlowck replied:
Thanks @jhaals ... that's true. Since we are on this.. if we are only creating a single worker, do we even need the `TaskWorker` or can we just use the `TaskBroker` to orchestrate the WorkflowRunners directly?

howlowck replied:
I removed the workers array, and the router is always just creating a single worker.

benjdlambert replied:
I think that we probably want to keep this as 3 right? Otherwise isn't this going to degrade performance without switching to the new property?

benjdlambert replied:
Or maybe it doesn't actually matter as it's going to use the other method anyway.

jhaals commented on plugins/scaffolder-backend/src/ScaffolderPlugin.ts:
```diff
@@ -36,7 +36,6 @@ import { createRouter } from './service/router';
  */
 export type ScaffolderPluginOptions = {
   actions?: TemplateAction<any>[];
-  taskWorkers?: number;
```
We need to keep this option as it would be a breaking change and require a major release if we removed it.
I think what we should do is default it to one and deprecate this option.

howlowck replied:
I added this option back. Thanks!

jhaals commented on plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:
```diff
@@ -49,6 +51,7 @@ export type CreateWorkerOptions = {
   workingDirectory: string;
   logger: Logger;
   additionalTemplateFilters?: Record<string, TemplateFilter>;
+  concurrentTasksLimit?: number;
```
Some docs stating the default 10 and that `Infinity` is the thing you should set if you don't want any limit

howlowck replied:
Thanks @jhaals, which file should this docs content be added? Or do you mean inline as JSDocs?

howlowck replied:
<img width="417" alt="image" src="https://user-images.githubusercontent.com/338265/196295202-15e76430-20a2-4256-9ec1-c223445be9bd.png">

jhaals reviewed (changes requested):
Thanks for the tweaks :) I've added some additional notes.

Diagrams is nice but let's make sure that they're accurate of the current implementation and used in the documentation if we intend to keep them part of this PR

Rugvip commented on plugins/scaffolder-backend/src/service/router.ts:
```diff
@@ -227,7 +211,17 @@ export async function createRouter(
   actionsToRegister.forEach(action => actionRegistry.register(action));
-  workers.forEach(worker => worker.start());
+
+  const worker = await TaskWorker.create({
```
@punkle this doesn't impact your setup does it? I'm thinking you'll also gladly have the same switch from multiple workers to a single worker with capacity for multiple jobs.

punkle replied:
Will this create a worker to run with the API? Is it possible to have a switch to turn this off? We dont want scaffolder templates to run along side the API process?

Rugvip replied:
Ah right, that's the customization you needed. I guess the simplest route is the one we're already at, i.e. if `concurrentTasksLimit` is set to `0`, then we won't bother creating a worker to execute tasks.

howlowck replied:
I reverted the changes I made regarding multiple TaskWorker creation. Now the `router` file stays largely unchanged (except the default TaskWorker counter is back to 1). Maybe the work of coalescing the `concurrentTasksLimit` and `taskWorkers` option can be done in another PR.

Rugvip commented on plugins/scaffolder-backend/src/service/router.ts:
```diff
@@ -67,7 +67,6 @@ export interface RouterOptions {
   scheduler?: PluginTaskScheduler;
 
   actions?: TemplateAction<any>[];
-  taskWorkers?: number;
```
We can't remove this since that's a breaking change. What I'm thinking we do instead is that we expose `concurrentTasksLimit` as an option here as well, and we make the two synonymous. Basically this:

```ts
const concurrentTasksLimit = options.concurrentTasksLimit ?? options.taskWorkers
```

We then deprecate `taskWorkers`, so that we can remove it once we hit 2.0 of `scaffolder-backend`

howlowck replied:
I added taskWorkers option back.

Rugvip commented on plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:
```diff
@@ -83,14 +91,15 @@ export class TaskWorker {
       for (;;) {
         const task = await this.options.taskBroker.claim();
-        await this.runOneTask(task);
+        this.taskQueue.add(() => this.runOneTask(task));
```
We're not `await`ing anything here, which we'll need to do. We need something where we wait if the queue is full, but otherwise we submit the task and move on to the next one.

howlowck replied:
we cannot await this `taskQueue.add` method call. Through experiementation, by `await` the `add` method would effectively make the `TaskWorker` an one-by-one blocking task execution again.

The mental model of the current code is, we wait for new tasks with `claim` (so it's not an infinite loop), then we just add them to the queue without limit nor waiting, and it's up to the queue to orchestrate the starting of the tasks with the configured `concurrentTasksLimit`.

Rugvip replied:
We don't want to be claiming tasks that we can't process yet though

howlowck replied:
Gotcha! Thanks for the clarification. I will use the [`empty` event on PQueue](https://github.com/sindresorhus/p-queue#empty) to claim only when the queue is empty.

howlowck replied:
... Actually we can't use `empty` because it's [only available in 7.3.0 of p-queue](https://github.com/sindresorhus/p-queue/releases/tag/v7.3.0), and we can't use ^7.0 because [it's ESM-only](https://github.com/sindresorhus/p-queue#install). I'll look into other options.

howlowck replied:
I added waiting `onEmpty` before TaskWorkers tries to claim new Tasks, so the Worker will block and only continue if the queue is empty.

Rugvip reviewed (changes requested):
Thanks! Couple of comments.

I'm thinking we can remove the second drawio diagram as well right? The first once could make sense to keep, but probably move it over into the scaffolder-backend itself or something like that? Alternatively find a place to reference it in the software templates docs

jhaals commented on plugins/scaffolder-backend/src/scaffolder/actions/builtin/stress/waitWorkflow.ts:
```diff
@@ -0,0 +1,102 @@
+
+import { createTemplateAction } from '../../createTemplateAction';
+
+export function waitWorkflow() {
```
I'm not sure that we want to expose the stress template and this action to users? It does not provide any usefulness more than testing the wait logic right?

It would be great if we could add some unit tests for the TaskWorker to ensure that this new behaviour work as expected :)

howlowck replied:
Tests Added :)

Rugvip commented on plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:
```diff
@@ -83,14 +103,16 @@ export class TaskWorker {
   start() {
     (async () => {
       for (;;) {
+        await this.taskQueue.onEmpty();
```
I'm assuming onEmpty waits for the queue to be completely empty? In that case we'll only process one task at a time. I think to jhaal's point it's best to use a tdd approach here and set up unit tests for the behavior that we want first.

howlowck replied:
@Rugvip onEmpty awaits until the queue is empty, ie. no tasks are waiting to be started. It is not waiting for all tasks are done. I have been validating the funcationality with the long-running action while developing the feature, and I know `concurrentTasksLimit` is working as expected. But I will include tests to demostrate the functionality this week.

Rugvip replied:
Ah alright, makes sense but not what I would've expected that to do xP

This'll still pick up tasks before we're actually ready to process them though right?

howlowck replied:
Hi @Rugvip, it will pick up **one** task when the waiting queue is empty.

howlowck replied:
I added tests to show the concurrent tasks on a single worker.

jhaals reviewed (changes requested):
We'd also want to deprecate the `taskWorkers` option from `RouterOptions` right?

It's worth mentioning that the amount of workers have been reduced to 1 but that we now have this new concurrency option that you can configure. Is `concurrentTasksLimit` exposed all the way up currently? I don't see where a user having their scaffolder configured as in https://github.com/backstage/backstage/blob/master/packages/backend/src/plugins/scaffolder.ts would be able to set this new option.

howlowck commented:
@jhaals Thanks for the feedback regarding exposing the `concurrentTasksLimit` option and deprecating the `taskWorkers` option. I plan to those changes this week.

Quick question, once the `concurrentTasksLimit` is exposed in the `RouterOptions`, do I add it in the [/backend/src/plugins/scaffolder.ts](https://github.com/backstage/backstage/blob/master/packages/backend/src/plugins/scaffolder.ts) file? (I don't see the existing `taskWorkers` option in there.) If so, what value do I set there to allow the user to pipe the value dynamically?

jhaals commented:
You're welcome! There is no need to add it to plugins/scaffolder.ts as it just used for development in this repository. We also want a sane default so it should typically not be changed unless you know what you're doing. It's however the main customisation point and an example of what the scaffolder.ts typically look like in a deployment. I assume this is an option that we want to expose from the "outside" as we have until now been exposing an option for taskWorkers.

howlowck commented:
Hi @jhaals and team, I updated the RouterOptions type and updated (regenerated) the `api-report.md`, but I'm getting failure on the "CI / verify" build. I would like to ask for some help me with the api-report generation process. I read the `REVIEWING.md` docs on the API Report, but I'm not sure how to properly generate the api report and get the `CI / verify` step to not fail. Thanks!

I ran `yarn build:api-reports:only plugins/scaffolder-backend` to generate the api-report since that's the only package I updated.

jhaals commented:
Hi @howlowck, I think you want to run `yarn build:api-reports plugins/scaffolder-backend` without the :only part. When I run it in your branch I do get an updated report.

howlowck commented:
Thanks @jhaals for your guidance. Not sure if this is an issue that the team is aware of but, for some reason it seems running the follow three commands all produced different results:
* `yarn build:api-reports:only plugins/scaffolder-backend`
* [`yarn build:api-reports plugins/scaffolder-backend`](https://github.com/backstage/backstage/pull/13817/commits/061f19dff1a3d5be8dfef1519aa77760eea69601)
* [`yarn build:api-reports`](https://github.com/backstage/backstage/pull/13817/commits/303c68d1f5fef8689c16fd26b8bc4edd16a91b2c) (this was the only one that passed the CI / Validate pipeline because this was the exact command that ran in the pipeline)

I think the PR is finally ready to go now. Thanks!

Rugvip commented on plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts:
```diff
@@ -127,3 +129,130 @@ describe('TaskWorker', () => {
+    new (options: TaskWorkerOptions): TaskWorker;
+  };
+
+  it('should not pick up tasks before it is ready to execute more work', async () => {
```
What I mentioned in https://github.com/backstage/backstage/pull/13817/files#r998666872 still seems to be an issue, so I added this failing test to highlight what I mean. Does this clarify my concern?

howlowck replied:
Thanks @Rugvip I changed how the queue claims tasks (only when the number of pending tasks is less than the limit will the worker claim a task). Your test now passes.

Rugvip commented on plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:
```diff
@@ -83,18 +103,33 @@ export class TaskWorker {
+    return new Promise(resolve => {
+      // "next" event emits when a task completes
+      // https://github.com/sindresorhus/p-queue#next
+      this.taskQueue.on('next', () => {
```
Think this needs a `once` or an unregister right? Think we'd be leaking listener's otherwise?

Rugvip reviewed (changes requested):
Nice! 👍 Think we're pretty much good now, just a tiny concern

jhaals commented:
Hoping we could push this over the finish line, feels like we're very close 🤞

howlowck commented:
Hi all! sorry. I'm out of the country at the moment. I will be back this weekend and make the requested update then. Thank you all for the support and the feedback thus far!

Rugvip reviewed (approved):
Nice, thank you! 👍 

Let's :shipit: 🎉

jhaals reviewed (approved):
👏  great work!
